tsdb/store.go

```diff
@@ -1124,23 +1124,42 @@ func (s *Store) DeleteMeasurement(database, name string) error {
 	})
 }
 
-// filterShards returns a slice of shards where fn returns true
+// filterShards takes in a list of functions and
+// returns a slice of shards where fn returns true
 // for the shard. If the provided predicate is nil then all shards are returned.
 // filterShards should be called under a lock.
-func (s *Store) filterShards(fn func(sh *Shard) bool) []*Shard {
+func (s *Store) filterShards(fns []func(sh *Shard) bool) []*Shard {
 	var shards []*Shard
-	if fn == nil {
+
+	// Early return if nil or 0 len
+	if len(fns) == 0 || fns == nil {
 		shards = make([]*Shard, 0, len(s.shards))
-		fn = func(*Shard) bool { return true }
-	} else {
-		shards = make([]*Shard, 0)
+		fn := func(*Shard) bool { return true }
+
+		for _, sh := range s.shards {
+			if fn(sh) {
+				shards = append(shards, sh)
+			}
+		}
+
+		return shards
```

Comment:
Maybe it's late in the day and I'm not comprehending something, but wouldn't all this be equivalent to:
```go
    copy(shards, s.shards)
    return shards
```

Reply:
Yes - this would be a much better way of doing it.

Reply:
unfortunately s.shards is of type `map[uint64]*Shard` so this is not possible. I can try and just do something simpler like:

```
		shards = make([]*Shard, 0, len(s.shards))
		for _, sh := range s.shards {
			shards = append(shards, sh)
		}
```

I do not believe go has a way to "get all values from a map" without looping or I would do it that way.